core/src/test/java/org/apache/calcite/test/MaterializedViewSubstitutionVisitorTest.java

```diff
@@ -1849,6 +1849,43 @@ protected final MaterializedViewFixture sql(String materialize,
     sql(mv2, query2).ok();
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-6193">[CALCITE-6193]
+   * If a query has more than one subexpression that matches a materialized view,
+   * only the first is substituted</a>. */
+  @Test void testStopTryIncorrectSubtree() {
+    final String mv = ""
+        + "select \"empid\", \"deptno\"\n"
+        + "from \"emps\"\n"
+        + "group by \"empid\", \"deptno\"";
+    final String matchQuery = "select \"deptno\"\n"
+        + "from \"emps\"\n"
+        + "group by \"deptno\"\n";
+    final String query = ""
+        + "select t1.\"deptno\"\n"
+        + "from (\n"
+        + "select \"deptno\"\n"
+        + "from \"emps\"\n"
+        + "union all\n"
+        + matchQuery
+        + ") as t1 inner join (\n"
+        + matchQuery
+        + ") as t2 on t1.\"deptno\" = t2.\"deptno\"\n";
+    sql(mv, query)
+        .checkingThatResultContains(""
+            + "LogicalCalc(expr#0..1=[{inputs}], deptno=[$t0])\n"
+            + "  LogicalJoin(condition=[=($0, $1)], joinType=[inner])\n"
+            + "    LogicalUnion(all=[true])\n"
+            + "      LogicalCalc(expr#0..1=[{inputs}], deptno=[$t1])\n"
```

Comment:
The 1879 and 1880 lines are still be replaced by mv's target descendant, it's parent may be shared with a wrong node, so I think this solution is not very is not very correct.

Reply:
If I don't commit, the RelNode is
```
LogicalCalc(expr#0..1=[{inputs}], deptno=[$t0])
  LogicalJoin(condition=[=($0, $1)], joinType=[inner])
    LogicalUnion(all=[true])
      LogicalCalc(expr#0..1=[{inputs}], deptno=[$t1]) -- same as 1879
        LogicalCalc(expr#0..4=[{inputs}], proj#0..1=[{exprs}]) -- same as 1880
          LogicalTableScan(table=[[hr, emps]])
      LogicalAggregate(group=[{1}])
        EnumerableTableScan(table=[[hr, MV0]])
    LogicalAggregate(group=[{0}])
      LogicalCalc(expr#0..4=[{inputs}], deptno=[$t1])
        LogicalTableScan(table=[[hr, emps]])
```
There are two `LogicalCalc`, do you mean this is Error?

Reply:
Yes, because these two LogicalCalc operators both are replaced by mv's target descendant, I think these temporary replacements should be undo after the real replacement has been done. This is the correct solution for this problem.